lib/Backend/arm64/LowerMD.cpp

```diff
@@ -1646,82 +1630,68 @@ LowererMD::LowerExitInstr(IR::ExitInstr * exitInstr)
+        for (RegNum reg = RegR8; reg <= LAST_CALLEE_SAVED_GP_REG; reg = (RegNum)(reg + 1))
         {
-            // We do need to deallocate the area allocated when we homed the params (since we weren't able to fold
-            // it into the first stack deallocation).
-            // TODO: Consider folding this into the LDM instruction above by having it restore dummy registers.
-
-            IR::RegOpnd * spOpnd = IR::RegOpnd::New(nullptr, this->GetRegStackPointer(), TyMachReg, this->m_func);
-            IR::IntConstOpnd * adjustOpnd = IR::IntConstOpnd::New(homedParams * MachRegInt, TyMachReg, this->m_func, true);
-            IR::Instr * spAdjustInstr = IR::Instr::New(Js::OpCode::ADD, spOpnd, spOpnd, adjustOpnd, this->m_func);
-            exitInstr->InsertBefore(spAdjustInstr);
+            if (usedRegs.Test(RegEncode[reg]))
+            {
+                regPopedSize += MachRegInt;
+                IR::Instr * instrStore = IR::Instr::New(Js::OpCode::LDR, this->m_func);
+                instrStore->SetSrc1(IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), regPopedSize, TyMachReg, this->m_func));
+                instrStore->SetDst(IR::RegOpnd::New(reg, TyMachReg, this->m_func));
+                exitInstr->InsertBefore(instrStore);
+            }
         }
-
-        // LR is still valid, so return by branching to it.
-        IR::Instr *  instrRet = IR::Instr::New(
-            Js::OpCode::RET,
-            IR::RegOpnd::New(nullptr, RegPC, TyMachReg, this->m_func),
-            IR::RegOpnd::New(nullptr, RegLR, TyMachReg, this->m_func),
-            this->m_func);
-        exitInstr->InsertBefore(instrRet);
     }
-    else
-    {
-        // 3. Set up original frame pointer - pop r11
-        usedRegs.ClearAll();
-        usedRegs.Set(RegEncode[RegR11]);
-        IR::Instr * instrPop = IR::Instr::New(Js::OpCode::POP, this->m_func);
-        instrPop->SetDst(IR::RegBVOpnd::New(usedRegs, TyMachReg, this->m_func));
-        instrPop->SetSrc1(IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP,TyMachReg, this->m_func), (int32)0, TyMachReg, this->m_func));
-        exitInstr->InsertBefore(instrPop);
 
-        // 4. Deallocate homed param area (if necessary) and return.
+    if (regPopedSize > 0)
+    {
+        GenerateStackDeallocation(exitInstr, regPopedSize);
+    }
 
-        // SP now points to the location where we saved LR.
-        // So return by doing a LDR pc,[sp],#n, where the postincrement of SP deallocates what remains of the stack.
-        // Note: the offset on this indir indicates the postincrement, which is the homed param area plus the size
-        // of LR itself.
-        IR::IndirOpnd * spIndir = IR::IndirOpnd::New(
-            IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func),
-             (homedParams + 1) * MachRegInt,
-            TyMachPtr, this->m_func);
+    // Generate LDP {fp,lr}
+    IR::Instr * instrStp = IR::Instr::New(Js::OpCode::LDP_POST, this->m_func);
```

Comment:
Again, LDP is limited to +512 byte offset, so you will need a fallback case for bigger allocations. TODO item for sure.

Reply:
Relative to SP? The offset here is only 16

---
In reply to: [143539590](https://github.com/Microsoft/ChakraCore/pull/3903#discussion_r143539590) [](ancestors = 143539590)

Reply:
Sorry you're right, it is only a fixed amount. My bad.